equihash/equihash.go

```diff
@@ -0,0 +1,662 @@
+
+	//  1) Generate list (X)
+	var X []hashKey
+	var tmpHash []byte
+	for i := 0; i < powOf2(collisionLength+1); i++ {
+		r := i % indicesPerHashOutput
+		if r == 0 {
+			currDigest := copyHash(digest)
+			err := hashXi(currDigest, i/indicesPerHashOutput)
+			if err != nil {
+				return nil, err
+			}
+			tmpHash = hashDigest(currDigest)
+		}
+		d := tmpHash[r*n/8 : (r+1)*n/8]
+		expanded, err := expandArray(d, hashLength, collisionLength, 0)
+		if err != nil {
+			return nil, errors.New("expandArray err: " + err.Error() + "\n")
+		}
+		X = append(X, hashKey{expanded, []int{i}})
+	}
+
+	// 3) Repeat step 2 until 2n/(k+1) bits remain
+	for i := 1; i < k; i++ {
+		// sort tuples by hash
+		sort.Sort(hashKeys(X))
+
+		var xc []hashKey
+		for len(X) > 0 {
+			// 2b) Find next set of unordered pairs with collisions on first n/(k+1) bits
+			xSize := len(X)
+			j := collisionOffset(X, i, collisionLength)
+
+			//2c) Store tuples (X_i ^ X_j, (i, j)) on the table
+			for l := 0; l < j-1; l++ {
+				for m := l + 1; m < l; m++ {
+					x1l, x1m := X[xSize-1-l], X[xSize-1-m]
+					if hasDistinctIndices(x1l.indices, x1m.indices) {
+						concat := concatIndices(x1l.indices, x1m.indices)
+						a, b := x1l.hash, x1m.hash
+						xc = append(xc, hashKey{xor(a, b), concat})
+					}
+				}
+			}
+			// 2d) drop this set
+			X = X[:len(X)-j]
+		}
+		// 2e) replace previous list with new list
+		X = xc
+	}
+
+	sort.Sort(hashKeys(X))
+	//find solutions
+	var solns []equihashSolution
+	for len(X) > 0 {
+		xn := len(X)
+		j := solutionOffset(X, k, collisionLength)
+		for l := 0; l < j-1; l++ {
+			for m := l + 1; m < j; m++ {
+				a, b := X[xn-1-l], X[xn-1-m]
+				res := xor(a.hash, b.hash)
+				f1 := countZeros(res) == 8*hashLength
+				f2 := hasDistinctIndices(a.indices, b.indices)
+				if f1 && f2 {
+					indices := concatIndices(a.indices, b.indices)
+					solns = append(solns, indices)
+				}
+			}
+		}
+		X = X[:len(X)-j]
+	}
+
+	return solns, nil
+}
+
+// countZeros counts leading zero bits in byte array
+func countZeros(h []byte) int {
+	for i, val := range h {
+		for j := 0; j < 8; j++ {
+			mask := 1 << uint(7-j)
+			if (int(val) & mask) > 0 {
+				return (i * 8) + j
+			}
+		}
+	}
+	return len(h) * 8
+}
+
+// returns the first index of second index that doesn't collide with the first hash
+func solutionOffset(x []hashKey, k, collisionLen int) int {
+	xSize := len(x)
+	for j := 1; j < xSize; j++ {
+		lc := hasCollision(x[xSize-1].hash, x[xSize-1-j].hash, k, collisionLen)
+		rc := hasCollision(x[xSize-1].hash, x[xSize-1-j].hash, k+1, collisionLen)
+		if !(lc && rc) {
+			return j
+		}
+	}
+	return xSize
+}
+
+// pow returns pow of base 2 for only positive k
+func powOf2(k int) int {
+	if k < 1 {
+		return 1
+	}
+	return 1 << uint(k)
+}
+
+// hasDuplicateIndices checks for duplicate indices within the same slice
+// returns true if there are duplicate numbers (indices)
+func hasDuplicateIndices(indices []int) bool {
+	if len(indices) <= 1 {
```

Comment:
we dont really need this `if`

Reply:
It's intentional because it avoids an allocation.

Reply:
For me, readability and simplicity is much more important than microoptimizations. But I get your point.